api/odilo.py

```diff
@@ -0,0 +1,532 @@
+        attached to this licensepool.
+
+        :param internal_format: Represents the patron's desired book format.
+
+        :return: a LoanInfo object.
+        """
+        record_id = licensepool.identifier.identifier
+
+        # Data just as 'x-www-form-urlencoded', no JSON
+        payload = dict(patronId=patron, format=internal_format)
+
+        response = self.patron_request(
+            patron, pin, self.CHECKOUT_ENDPOINT.replace("{recordId}", record_id),
+            extra_headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            data=payload)
+
+        response_json = response.json()
+        if response.status_code == 404:
+            if response_json and response_json['errors'] and len(response_json['errors']) > 0:
+                error = response_json['errors'][0]
+                if error['id'] == 'ERROR_DATA_NOT_FOUND':
+                    raise NotFoundOnRemote('Record %s' % record_id)
+
+                elif error['id'] == 'NoCopiesAvailable':
+                    # Clearly our info is out of date.
+                    # self.update_licensepool(identifier.identifier)
+                    raise NoAvailableCopies()
+                elif error['id'] == 'PatronHasExceededCheckoutLimit':
+                    raise PatronLoanLimitReached()
+                else:
+                    raise CannotLoan(error['id'])
+        else:
+            return self.loan_info_from_odilo_checkout(licensepool.collection, response_json)
+
+    def loan_info_from_odilo_checkout(self, collection, checkout):
+        start_date = self.extract_date(checkout, 'startTime')
+        end_date = self.extract_date(checkout, 'endTime')
+
+        return LoanInfo(
+            collection,
+            DataSource.ODILO,
+            Identifier.ODILO_ID,
+            checkout['id'],
+            start_date,
+            end_date,
+            checkout['downloadUrl']
+        )
+
+    def checkin(self, patron, pin, licensepool):
+        checkout_id = licensepool.identifier.identifier
+        url = self.CHECKIN_ENDPOINT.replace("{checkoutId}", checkout_id) + "?patronId=" + patron
+
+        response = self.patron_request(patron, pin, url, method='POST')
+        if response.status_code == 404:
+            data = response.json()
+            if data and data['errors'] and len(response.json()['errors']) > 0:
+                error = data['errors'][0]
+                if error['id'] == 'CHECKOUT_NOT_FOUND':
+                    raise NotCheckedOut(error['description'])
+                else:
+                    raise CannotReturn(error['id'] + ': ' + error['description'])
+        else:
+            return response
+
+    @classmethod
+    def extract_date(cls, data, field_name):
+        if field_name not in data:
+            d = None
+        else:
+            # OdiloAPI dates are timestamps in milliseconds
+            d = datetime.datetime.fromtimestamp(float(data[field_name]) / 1000.0)
```

Comment:
fromtimestamp converts a time to local time, and we use UTC exclusively. When we need to convert a timestamp to milliseconds, we create an 'epoch' datetime for January 1, 1970, and then add `datetime.timedelta(milliseconds=timestamp)` to it.

Reply:
ok, changed. What about using ```datetime.datetime.utcfromtimestamp``` ??

Reply:
Good catch. I think `utcfromtimestamp` will work fine.